viewing_party/party.py

```diff
@@ -1,23 +1,144 @@
+from collections import Counter
+
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    """Create a dictionary containing information about a movie."""
+    if not title or not genre or not rating:
+        return None
+    return {
+        "title": title,
+        "genre": genre,
+        "rating": rating,
+    }
+
+def add_to_watched(user_data, movie):
+    """Add a movie to a user's watched list."""
+    updated_data = user_data.copy()
+    updated_data["watched"].append(movie)
+    return updated_data
+
+def add_to_watchlist(user_data, movie):
+    """Add a movie to a user's watchlist."""
+    updated_data = user_data.copy()
+    updated_data["watchlist"].append(movie)
+    return updated_data
+
+def watch_movie(user_data, movie):
+    """If a movie is in the user's watchlist, add it to watched."""
+    updated_data = user_data.copy()
+    watchlist = updated_data["watchlist"]
+    watched = updated_data["watched"]
+    watchlist_titles = {
+        watch_item["title"]: watch_item 
+        for watch_item in watchlist
+    }
+    if movie in watchlist_titles:
+        movie_details = watchlist_titles[movie]
+        watchlist.remove(movie_details)
+        watched.append(movie_details)
+    return updated_data
+        
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    """Return the average rating of a user's watched movies. Return 0 if 
+    watched is empty."""
+    watched = user_data["watched"]
+    total = 0
+    for movie in watched:
+        total += movie["rating"]
+    return (total / len(watched)) if len(watched) else 0
+
+def get_most_watched_genre(user_data):
+    """Return the most watched genre in a user's watched list.""" 
+    genres = Counter([movie["genre"] for movie in user_data["watched"]])
+    return genres.most_common(1)[0][0] if genres else None
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    """Return all the movies the user has watched that their friends have not 
+    watched."""
+    user_watched = {movie["title"]: movie for movie in user_data["watched"]}
+    friends_watched = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched.add(movie["title"])
+    for title in friends_watched:
+        if title in user_watched:
+            user_watched.pop(title)
+    return list(user_watched.values())

```

Comment:
👍  Nice work! Here are a few thoughts:

- Consider making a helper function to return a 'flattened' list of all the movies the user's friends have watched. The act of 'flattening' is used to describe creating a new one-dimensional list based on a nested data structure. We will actually do this a lot, especially while working with APIs.
- What is the time and space complexity of this function? How could we improve it?